chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,637 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use static_assertions::assert_eq_size;
+use std::collections::BTreeMap;
+
+use crate::{error::ConnectTransactionError, TransactionSource};
+
+use common::{
+    chain::{Block, DelegationId, PoolId},
+    primitives::{amount::UnsignedIntType, Amount, Id},
+    Uint128, Uint256,
+};
+use pos_accounting::{
+    AccountingBlockRewardUndo, PoSAccountingOperations, PoSAccountingUndo, PoSAccountingView,
+    PoolData,
+};
+use utils::ensure;
+
+use super::accounting_delta_adapter::PoSAccountingDeltaAdapter;
+
+/// Distribute reward among the pool's owner and delegations
+pub fn distribute_pos_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_reward: Amount,
+) -> Result<AccountingBlockRewardUndo, ConnectTransactionError> {
+    let pool_data = accounting_adapter
+        .accounting_delta()
+        .get_pool_data(pool_id)?
+        .ok_or(ConnectTransactionError::PoolDataNotFound(pool_id))?;
+
+    let pool_owner_reward = calculate_pool_owner_reward(total_reward, &pool_data).ok_or(
+        ConnectTransactionError::PoolOwnerRewardCalculationFailed(block_id, pool_id),
+    )?;
+
+    let total_delegations_reward = (total_reward - pool_owner_reward).ok_or(
+        ConnectTransactionError::PoolOwnerRewardCannotExceedTotalReward(
+            block_id,
+            pool_id,
+            pool_owner_reward,
+            total_reward,
+        ),
+    )?;
```

Comment:
Do we really need `PoolOwnerRewardCannotExceedTotalReward` error? It can't happen, so maybe it's better to use `expect` instead?

Reply:
It's just paranoia. Getting an error that we can handle is better than crashing. Things can go wrong.